### ISSUE TITLE
Is there any hardware CRC instruction on Intel's processors to accelerate CRC in ISA-L?

### ISSUE THREAD

**Issue by YamboZhou** [intel/isa-l#170]
Hi guys,

We used hardware CRC instruction on Intel's SSE 4.2 processors before.  Now when we use crc16_t10 in ISA-L, we notice that this is a software-based CRC algorithm that costs a lot of CPU cycles.  

Is there any specific CPU architecture that could support hardware CRC instruction for crc16_t10 in ISA-L?

Thanks!

**Comment by YamboZhou** [comment]
BTW,  I notice that in the following paper, crc16_t10 in ISA-L could be accelerated by PCLMULQDQ instruction. 

https://www.intel.com/content/dam/www/public/us/en/documents/white-papers/fast-crc-computation-generic-polynomials-pclmulqdq-paper.pdf

Below is my machine's CPU information and flags:

`Intel(R) Xeon(R) CPU E5-2682 v4 @ 2.50GHz

flags   : fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush dts acpi mmx fxsr sse sse2 ss ht tm pbe syscall nx pdpe1gb rdtscp lm constant_tsc arch_perfmon pebs bts rep_good nopl xtopology nonstop_tsc aperfmperf eagerfpu pni **pclmulqdq** dtes64 ds_cpl vmx smx est tm2 ssse3 fma cx16 xtpr pdcm pcid dca sse4_1 sse4_2 x2apic movbe popcnt tsc_deadline_timer aes xsave avx f16c rdrand lahf_lm abm 3dnowprefetch ida arat epb pln pts dtherm tpr_shadow vnmi flexpriority ept vpid fsgsbase tsc_adjust bmi1 hle avx2 smep bmi2 erms invpcid rtm cqm rdseed adx smap xsaveopt cqm_llc cqm_occup_llc`

And there has PCLMULQDQ  flag in our machine. I use function **crc16_t10dif_base** to make CRC computation in our application. But it looks like it still use software routines that don't use  hardware instruction. 

I am wondering whether there is any wrong configuration in our environment or crc16_t10 in ISA-L can not be accelerated by hardware instruction?

Thanks for your help and time!

**Comment by jeffareid** [comment]
There are 16 bit CRC code examples that are nearly identical to 32 bit CRC code examples, using the upper bits, then shifting right 16 bits when done, such as this one:

https://github.com/intel/isa-l/blob/master/crc/crc16_t10dif_02.asm

**Comment by YamboZhou** [comment]
Thanks @jeffareid, it works.

**Comment by gbtucker** [comment]
@YamboZhou, I would suggest calling the main multi-binary interface crc16_t10dif() not the base function crc16_t10dif_base().  The former will choose between five different versions based on your machine.

**Comment by YamboZhou** [comment]
@gbtucker, thanks. I'm now using multi-binary interface crc16_t10_dif(). It works. The CPU consumption is less than 3%.

**Comment by aakefbs** [comment]
I'm just looking for the fastest CRC and found this discussion here. @YamboZhou I actually wonder why you are using crc16, when you had been using crc32c before? "make perf" reports ~15.5GB/s for crc32_iscsi on my system (Haswell E3-1245 v3),  but only ~12.7GB/s for crc16_t10dif.

Hmm, interesting, the above is my workstation, the real system where fast a crc is a Haswell E5-2695 v3 and there it is

crc/crc16_t10dif_perf: 11650.12 MB/s
crc/crc32_iscsi_perf: 12016.63 MB/s

I guess this is due to the higher base  frequency of my workstation and as the test is only single threaded. Now what is the advantage of crc16 to crc32? Less space to store it? 
In my case,  I guess I need a multi-threaded benchmark, to see which is the fastest with that.

**Comment by jeffareid** [comment]
Intel processors with SSE4.2 include a CRC32C instruction, which is faster than the 500+ line assembly code that handles generic CRC for 16 | 32 | 64 bit CRC.  The times for the assembly code are about the same for 16 | 32 | 64, most of the code is the same, with just a different set of constants, only 2 or 3 additional instructions for 16 | 64 bit CRC.

**Comment by aakefbs** [comment]
Yeah, we are actually using that, but I wasn't sure if there are not any other faster AVX ways in the mean time. Ah, see that our code is almost the same as htot/crc32c/crc_iscsi_v_pcl.asm, which seems to come directly from intel.

What is the reason that the isal-l project does not have crc32c?

**Comment by gbtucker** [comment]
The ISA-L function crc32_iscsi() is the same polynomial and parameters that is often called crc32c, Castagnoli or others.  In some implementations of this function, it does use the special intel crc32 instruction, in others it is faster to use vclmulqdq.

**Comment by aakefbs** [comment]
Ah thanks a lot! And looking into crc32_iscsi_01.asm I also see the crc32 instrunction.
